text/0000-alternate-registries.md

```diff
@@ -0,0 +1,120 @@
+serde_json = "1.0.1"
+```
+
+## Changes for alternative registries for dependencies
+This boils down to a very simple change, where we previously setup the crate source for the
+crates.io registry, we now just need to check if a registry is provided, if it has the crate
+source is created using the registry URL, otherwise the crates.io server is used.
+
+## Blocking requests to push to a registry
+There are two parts to this, the first is a change to Cargo which checks if the registry
+provided in the registry matches the host for the publish, if it does not it gets rejected.
+The second part is a change to crates.io which will just reject the request to publish the
+crate if the configured repository on the crates.io server does not match the registry
+specified in the package or dependencies.
+
+## Allow publishing when referencing external dependencies
+We still want to support private crates having dependencies on the public crates.io server,
+so we propose relaxing a check which ensures that the source for a dependency matches the
+registry. We propose that this only performs the check only if the dependency is not the
+default registry, thus allowing private crates to reference public crates on crates.io.
+
+## Making it easier for users using an alternate crates.io registry
+When a user selects a specific crate the Cargo.toml fragment would be updated to include the
+registry URL, thus allowing users to easily copy and paste into their projects Cargo.toml
+file.
+
+# How We Teach This
+[how-we-teach-this]: #how-we-teach-this
+
+The term alternative registry would seem the most appropriate to describe this feature.
+
+In the first instance I think that the Cargo.toml format documentation is sufficient to
+provide access to the feature. However, in time once more of the pieces fall into place
+it would be useful having a guide on how to setup/administer a crates.io server in an
+enterprise setting (similar to the initial mirroring documentation).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Currently this design requires that when you want to push to the private crates.io server
+you need to override the host and token, it would be possible to update cargo to support
+multiple registries tokens which can be used to login.
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Using a single server for cache and private registry
+It was considered proposing a single crates.io server which performs both caching of crates.io,
+plus has the ability to have crates pushed to it, however this has the following drawbacks:
+* It requires crates.io to be able to combine two registries, or requires a radical change to the way crates.io works
+* The current proposal could be extended to support this, if a caching server is added at a later stage
+
+## Including registry definitions in a global location
+We considered using a global configuration file (eg ~/.cargo/config) to allow a registry to
```

Comment:
[Cargo will look up the directory tree for `.cargo/config` files](http://doc.crates.io/config.html#hierarchical-structure), so it is possible to have, for example, `~/some/path/my-project/.cargo/config` rather than globally.

I think Cargo.toml is a better place for this information since it's intended to be checked in and shared while `.cargo/config` files are meant to be ignored and personal, but I wanted to make sure you knew about `.cargo/config` being an option.